### PR TITLE
Stackdriver: Support meta labels

### LINKED ISSUE #18887: Stackdriver: Add support for filtering, grouping and alias by on metadata labels
As of now, the Stackdriver datasource supports filtering, grouping and using alias by on resource labels and metric labels. However the [stackdriver metrics API](https://cloud.google.com/monitoring/api/ref_v3/rest/v3/TimeSeries) also includes `metadata` which is an object that also includes labels - `user_labels` and `system_labels`. These labels might be useful in many cases, particularly when using agent metrics since they don't include instance_name as a metric label. 

It should be possible to use`metadata.user_labels` and `metadata.system_labels` for filtering, grouping and alias by. This will involve both backend and frontend work in the stackdriver datasource. 

Fixes #15552 
Fixes #18750

### PR DESCRIPTION
**Pull request by sunker**
This PR makes it possible to groupby, filter and alias on metadata labels. See detailed description in #18887. It also does a lot of refactoring in backend and in the frontend, e.g rewriting the group by and filter component from angular to react using the segment component. 

fixes #18887

### CODE REVIEW

File: pkg/tsdb/stackdriver/stackdriver.go
Comment (cbandy) [inline pkg/tsdb/stackdriver/stackdriver.go]: These `!exist` checks and assignments appear redundant. The following `containsLabel()` returns `false` either way, and `append()` will allocate as-needed.

For example, https://play.golang.org/p/qjovfSISn2U

File: pkg/tsdb/stackdriver/stackdriver.go
Comment (igorpeshansky) [inline pkg/tsdb/stackdriver/stackdriver.go]: Did this come from [here](https://www.golangprograms.com/golang-convert-string-into-snake-case.html)? I think that code guards against the initial capital letter and against runs of capital letters, which shouldn't happen in the API JSON response. Just curious, why wouldn't the following have sufficed?
```go
func toSnakeCase(str string) string {
	matchAllCap := regexp.MustCompile("(.)([A-Z][a-z]*)")
	return strings.ToLower(matchAllCap.ReplaceAllString(str, "${1}_${2}"))
}
```

**Review by igorpeshansky — CHANGES_REQUESTED** [review]
Also tagging @bmoyles0117 for more Go expertise.

File: pkg/tsdb/stackdriver/stackdriver.go
Comment (sunker) [reply pkg/tsdb/stackdriver/stackdriver.go]: Yes, of course. Thanks for feedback!

File: pkg/tsdb/stackdriver/types.go
Comment (sunker) [reply pkg/tsdb/stackdriver/types.go]: Okay, interesting! The json response is camel cased, but I'll change json definition to be lowercase (it works too).

File: pkg/tsdb/stackdriver/stackdriver.go
Comment (sunker) [reply pkg/tsdb/stackdriver/stackdriver.go]: That's is sufficient, thank you!

File: pkg/tsdb/stackdriver/types.go
Comment (sunker) [reply pkg/tsdb/stackdriver/types.go]: 👍

**Comment by sunker** [comment]
Thanks for feedback @igorpeshansky, I've made a few fixes and pushed them. Did you get at chance to test it too?

File: pkg/tsdb/stackdriver/stackdriver.go
Comment (bmoyles0117) [inline pkg/tsdb/stackdriver/stackdriver.go]: Instead of passing an empty map, you could just pass `nil`. `formatLegendKeys` does not mutate the map, reading from a nil map will not produce a panic.

**Review by igorpeshansky — APPROVED** [review]
LGTM :shipit:, modulo some minor remaining feedback. As I mentioned in #18887, I'm not really equipped to test it, but @joyw-2018 should be able to help with that (as well as make authoritative calls on the expected behavior).

File: pkg/tsdb/stackdriver/stackdriver.go
Comment (bmoyles0117) [inline pkg/tsdb/stackdriver/stackdriver.go]: This could be global, to avoid instantiating the regex object every time.

File: pkg/tsdb/stackdriver/stackdriver.go
Comment (sunker) [reply pkg/tsdb/stackdriver/stackdriver.go]: No that will just append the value to the output array. [This](https://github.com/grafana/grafana/blob/acb61a6a3561d346f7e322834864306e8d5ad5a4/pkg/tsdb/stackdriver/stackdriver_test.go#L434-L442) unit test should validate that behavior.

Comment (igorpeshansky) [reply pkg/tsdb/stackdriver/stackdriver.go]: For `labels[key]`, yes, which is why I didn't comment on that loop. However, for `seriesLabels[key]`, which becomes part of the legend, it looks like only the last value will be recorded....

File: pkg/tsdb/stackdriver/stackdriver.go
Comment (sunker) [reply pkg/tsdb/stackdriver/stackdriver.go]: Tried that intially, but it doesn't work. I guess go does only recognize at a slice are not aware of element types when casting. Anyway, refactored the code to use type switching instead of if statements. That should make it a bit more readable.

File: pkg/tsdb/stackdriver/stackdriver.go
Comment (sunker) [reply pkg/tsdb/stackdriver/stackdriver.go]: Yes, thank you!

File: pkg/tsdb/stackdriver/stackdriver.go
Comment (sunker) [reply pkg/tsdb/stackdriver/stackdriver.go]: 👍

File: pkg/tsdb/stackdriver/stackdriver.go
Comment (sunker) [reply pkg/tsdb/stackdriver/stackdriver.go]: Yes, should be good for performance. Thanks!

File: pkg/tsdb/stackdriver/stackdriver.go
Comment (bmoyles0117) [reply pkg/tsdb/stackdriver/stackdriver.go]: LGTM

**Review by bmoyles0117 — APPROVED** [review]
LGTM

File: pkg/tsdb/stackdriver/stackdriver.go
Comment (sunker) [reply pkg/tsdb/stackdriver/stackdriver.go]: Ah I see, good catch. Did some more refactoring. In case metadata.system_labels.x is a string array the values will be joined with a comma. 

`metadata.system_labels.name = ["value1", "value2"]`
when used in the alias by field like this
`{{metadata.system_labels.name}}` 
the legend will be named `value1, value2`

**Comment by sunker** [comment]
Thanks for feedback @igorpeshansky and @bmoyles0117. I'd like @daniellee to take a look at this too before merging. We'll aim for version 6.6 which will be released in beta state next week.

**Comment by ryantxu** [comment]
I don't know if this is a HUGE change or not... if possible, it would be great to have all the metadata keys live under: [meta.custom](https://github.com/grafana/grafana/blob/master/packages/grafana-data/src/types/data.ts#L28)

Currently the meta object allows any key/value *and* has some common ones.  This makes it difficult to have a well typed SDK with consistent behavior.  See: https://github.com/grafana/grafana/issues/21429  and  https://github.com/grafana/grafana-plugin-sdk-go/blob/0929c2dbb3a2ae13aabbfc474e5775eb90e815f3/dataframe/query_result_meta.go#L9

If this is a big change, then we can do it after 6.6

**Comment by sunker** [comment]
Good idea @ryantxu! Shouldn't be that much work, but I have a few other things I need to be focusing on right now so I'll get back to this in a week or two. Thanks!

File: pkg/tsdb/stackdriver/stackdriver.go
Comment (abhik) [inline pkg/tsdb/stackdriver/stackdriver.go]: I'm new to both grafana and stackdriver, so I may be missing something... but I believe some metrics are now named like `kubernetes.io/*`. See https://cloud.google.com/monitoring/kubernetes-engine/migration.

Comment (sunker) [reply pkg/tsdb/stackdriver/stackdriver.go]: Thanks for feedback @abhik! I've updated the regex to include kubernetes

**Review by marefr — APPROVED** [review]
Code looks okay. Lets get this merged for beta1.

**Pull request merged**
